### ISSUE TITLE
Fix edge case in constraint solver result caching

### ISSUE THREAD

**Issue by martijnwalraven** [meteor/meteor#6541]
This issue first came to light by investigating a failing self-test, `change packages during hot code push` in `package-tests.js`. It took a long time to unravel and can only be reproduced as a result of a fairly specific set of steps:
1. Add a local package to a project
2. Make a change to the package that does not affect the list of dependencies (e.g. change the contents of a file)
3. Add an `api.use` to the `package.js` referencing a local package that wasn't previously in the project

This results in an exception: `Error: Depending on unknown package <newly-included-package>`.

I tracked this down to the following circumstances:
- We cache previous constraint solver output and use it if the input hasn't changed (https://github.com/meteor/meteor/commit/ed822326aab0d88a648f4a8ff9f7246691be2a3c).
- We used to compare the complete contents of the constraint solver input, but because that includes `catalogCache` (which is populated as needed) they never seem to have matched. This was recently changed (https://github.com/meteor/meteor/commit/0376b477a54f8b90f5ee7c35c8043ec238c6b8c6). That explains why the issue wasn't encountered before.
- We do compare `previousSolution` however, hence the need for step 2. Because we make a change that doesn't affect dependencies there, `previousSolution` will be the same at step 3.
- That means we will reuse the previously cached output as long as the `dependencies` and `constraints` are the same.
- Because we [always add all local packages to the list of constraints](https://github.com/meteor/meteor/blob/release-1.3/tools/project-context.js#L644-L654), adding a new `api.use` referencing a local package doesn't affect `constraints`.
- So we are now in a situation where we reuse the existing output and end up with a `packageMap` that doesn't include the local package we depend on. This then leads to an exception [when processing the `api.use` call in `package-source.js`](https://github.com/meteor/meteor/blob/release-1.3/tools/isobuild/package-source.js#L1588-L1593).

I'm not sure what to do about this, and it might not be worth fixing now. The main problem is that constraint solver input doesn't fully determine constraint solver output, but result caching does assume that to be the case.

**Comment by tmeasday** [comment]
Actually you don't need to do step 2.

**Comment by tmeasday** [comment]
Actually, let me rephrase this. You can reproduce with the following:

1. Use a local package in your app
2. Run `meteor`
3. `api.use` a non-app package in the local package.

**Comment by tmeasday** [comment]
I removed the relevant section of the test. We should re-add it if we fix this bug

**Comment by hwillson** [comment]
I've confirmed this is still an issue with Meteor 1.6 (the test that was removed in https://github.com/meteor/meteor/commit/bd8ecfeec9c49bb10e5f5b8c0d914859bd6cbf92 still fails).

**Comment by hwillson** [comment]
Closing for now since this is definitely an edge case and there hasn't been any activity on this issue in quite some time. If anyone is encountering this issue and feels like it should be addressed, please note that this issue has been marked as `pull-requests-encouraged`. If you're interested in working on it, please let us know and we'll re-open. Thanks!
